src/lib/tls/msg_client_hello.cpp

```diff
@@ -463,6 +463,24 @@ Client_Hello_12::Client_Hello_12(std::unique_ptr<Client_Hello_Internal> data)
          m_data->extensions().add(new Renegotiation_Extension());
          }
       }
+   // RFC 8446 4.2.3
+   //    The client MUST NOT include these extensions [Supported Elliptic Curves Extension
+   //    and Supported Point Formats Extension] in the ClientHello message if it does not
+   //    propose any ECC cipher suites.
+   if(m_data->extensions().has<Supported_Groups>() || m_data->extensions().has<Supported_Point_Formats>())
+      {
+      auto offered_ciphers = m_data->ciphersuites();
+      if (std::none_of(offered_ciphers.begin(), offered_ciphers.end(), [](uint16_t cipherSuiteCode) {
+         std::optional<Ciphersuite> cs = Ciphersuite::by_id(cipherSuiteCode);
+         // true for ECC cipher suites or TLS 1.3 cipher suites
+         return cs.has_value() && (cs.value().ecc_ciphersuite() ||
+                                   cs.value().auth_method() == Auth_Method::UNDEFINED ||
+                                   cs.value().kex_method() == Kex_Algo::UNDEFINED); }))
```

Comment:
Maybe simplify to:

```suggestion
         return cs.has_value() && (cs.value().ecc_ciphersuite() ||
                                   cs.usable_in_version(Protocol_Version::TLS_V13)); }))
```

Or even add a predicate method to `Ciphersuite`, like:

```bool Ciphersuite::is_tls12_suite()```